src/System.Security.Cryptography.Cng/src/Internal/Cryptography/CngAlgorithmCore.cs

```diff
@@ -18,14 +19,40 @@ public static CngKey Duplicate(CngKey key)
             return CngKey.Open(key.Handle, key.IsEphemeral ? CngKeyHandleOpenOptions.EphemeralKey : CngKeyHandleOpenOptions.None);
         }
 
-        public CngKey GetOrGenerateKey(int keySize, CngAlgorithm algorithm)
+        public bool IsKeyGenerated
         {
-            // If our key size was changed from the key we're using, we need to generate a new key.
-            if (_lazyKey != null && _lazyKey.KeySize != keySize)
+            get
+            {
+                return (_lazyKey != null);
+            }
+        }
+
+        public bool IsKeyGeneratedNamedCurve()
+        {
+            return (_lazyKey != null && _lazyKey.IsECNamedCurve());
+        }
+
+        public void DisposeKey()
+        {
+            if (_lazyKey != null)
             {
                 _lazyKey.Dispose();
                 _lazyKey = null;
             }
+        }
+
+        public CngKey GetOrGenerateKey(int keySize, CngAlgorithm algorithm, ECCurve? curve = null)
```

Comment:
Makes sense. There wasn't actually that much shared code so it was more straightforward just to create two public methods